File: src/main/java/com/phosa/json/model/JsonObject.java
package com.phosa.json.model;

import com.alibaba.fastjson2.JSON;
import com.alibaba.fastjson2.JSONObject;
import com.phosa.json.exception.JsonException;

import java.util.Arrays;
import java.util.Map;
import java.util.Set;

public class JsonObject {
    JSONObject jsonObject;
    public JsonObject() {
        jsonObject = new JSONObject();
    }
    public JsonObject(JSONObject jsonObject) {
        this.jsonObject = jsonObject;
    }

    public int size() {
        return jsonObject.size();
    }

    public boolean containsKey(String key) {
        return jsonObject.containsKey(key);
    }

    public Object put(String key, Object value) {

        if (value instanceof JsonObject) {
            return this.jsonObject.put(key, ((JsonObject) value).getJSONObject());
        } else if (value instanceof JsonArray) {
            return this.jsonObject.put(key, ((JsonArray) value).getJSONArray());
        }
        return jsonObject.put(key, value);
    }

    public JsonObject fluentPut(String key, Object value) {
        put(key, value);
        return this;
    }

    public Object get(String key) {
        return jsonObject.get(key);
    }

    public String getAsString(String key) {
        return jsonObject.getString(key);
    }

    public JsonObject getAsJsonObject(String key) {
        return new JsonObject(jsonObject.getJSONObject(key));
    }
    public JsonArray getAsJsonArray(String key) {
        return new JsonArray(jsonObject.getJSONArray(key));
    }
    public boolean getAsBoolean(String key) {
        return jsonObject.getBoolean(key);
    }
    public int getAsInt(String key) {
        return jsonObject.getIntValue(key);
    }
    public double getAsDouble(String key) {
        return jsonObject.getDoubleValue(key);
    }
    public long getAsLong(String key) {
        return jsonObject.getLongValue(key);
    }
    public Set<Map.Entry<String, Object>> getEntrySet() {
        return jsonObject.entrySet();
    }
    public boolean containsValue(Object value) {
        return jsonObject.containsValue(value);
    }
    public boolean containsKey(Object key) {
        return jsonObject.containsKey(key);
    }
    public boolean containsAll(JsonObject jsonObject) {
        return this.getEntrySet().containsAll(jsonObject.getEntrySet());
    }
    public boolean containsAll(Object[] keys) {
        for (Object key : keys) {
            if (!jsonObject.containsKey(key)) {
                return false;
            }
        }
        return true;
    }
    public boolean containsAny(JsonObject jsonObject) {
        for (Object key : jsonObject.jsonObject.keySet()) {
            if (this.jsonObject.containsKey(key)) {
                return true;
            }
        }
        return false;
    }
    public boolean containsAny(Object[] keys) {
        for (Object key : keys) {
            if (jsonObject.containsKey(key)) {
                return true;
            }
        }
        return false;
    }
    public String getNestedValue(String... keys) {
        JsonObject current = this;
        for (String k : keys) {
            if (current != null) {
                current = current.getAsJsonObject(k);
            } else {
                throw new JsonException("Invalid key path: " + Arrays.toString(keys));
            }
        }
        return current != null ? current.toString() : "";
    }
    public String toString() {
        return JSON.toJSONString(this.jsonObject);
    }

    public JSONObject getJSONObject() {
        return jsonObject;
    }
}

File: src/main/java/com/phosa/json/model/JsonObject.kt
package com.phosa.json.model

import com.phosa.json.JsonUtil

class JsonObject(private var map: Map<String, Any>) {

    constructor() : this(emptyMap())

    fun getMap(): Map<String, Any> {
        return map
    }

    fun getAsBoolean(key: String): Boolean? {
        return get<Boolean>(key)
    }
    fun getAsInt(key: String): Int? {
        return get<Int>(key)
    }
    fun getAsString(key: String): String? {
        return get<String>(key)
    }
    fun getAsJsonObject(key: String): JsonObject? {
        return JsonObject(get<Map<String, Any>>(key)?:emptyMap())
    }

    fun getAsJsonArray(key: String): JsonArray? {
        return JsonArray(get<List<Any>>(key)?:emptyList())
    }

    inline fun <reified T> get(key: String): T? {
        return getMap()[key]?.takeIf { it is T }?.let { it as T }
    }

    fun containsKey(key: String): Boolean {
        return getMap().containsKey(key)
    }
    fun containsValue(value: Any): Boolean {
        return getMap().containsValue(value)
    }

    // 添加或更新一个键值对
    fun put(key: String, value: Any): Any? {
        val oldValue = map[key]
        val updatedMap = map.toMutableMap().apply {
            this[key] = when (value) {
                is JsonObject -> value.getMap()
                is JsonArray -> value.getList()
                else -> value
            }
        }
        map = updatedMap.toMap()
        return oldValue
    }

    // 扩展方法，用于打印 map 内容（可选）
    override fun toString(): String {
        return JsonUtil.toJson(getMap())
    }


}
